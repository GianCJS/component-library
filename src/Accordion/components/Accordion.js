import React, { useState } from 'react';
import PropTypes from 'prop-types';

function Accordion(props) {
    const { header, content, open, className, disabled, card } = props;
    const [accordionOpen, setAccordionOpen] = useState(open);

    function handleAccordionHeaderClick() {
        if (!disabled) {
            setAccordionOpen(!accordionOpen);
        }
    }

    const accordionClasses = `accordion-main-wrapper
        ${accordionOpen ? 'accordion-open' : 'accordion-closed'}
        ${disabled ? 'accordion-disabled accordion-closed' : ''}
        ${card ? 'accordion-card' : ''}
        ${className}
    `;
    return (
        <div className={accordionClasses}>
            <div className="accordion-header-wrapper" onClick={handleAccordionHeaderClick}>
                <div className="accordion-header-title">
                    {header}
                </div>
                <div className="accordion-header-caret-wrapper">
                    <span className="accordion-header-caret" />
                </div>
            </div>
            <div className="accordion-content-wrapper">
                <div className="accordion-content">
                    {content}
                </div>
            </div>
        </div>
    );
}

Accordion.defaultProps = {
    header    : '',
    content   : '',
    open      : false,
    className : '',
    disabled  : false,
    card      : true,
};

Accordion.propTypes = {
    /**
    Accordion header content
    */
    header    : PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
    /**
    Accordion content
    */
    content   : PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
    /**
    Set open state
    */
    open      : PropTypes.bool,
    /**
    Custom className
    */
    className : PropTypes.string,
    /**
    Set disabled state
    */
    disabled  : PropTypes.bool,
    /**
    Show as a card
    */
    card      : PropTypes.bool,
};

export default Accordion;
