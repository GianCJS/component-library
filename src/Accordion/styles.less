@accordion-padding: @space-normal;

/* Main styles */

.accordion-main-wrapper {
    border-radius: @base-border-radius;    

    & + .accordion-main-wrapper {
        margin-top: @space-small;
    }

    &.accordion-card {
        background: @card-background;

        &.accordion-open {
            box-shadow: @card-shadow;
        }

    }

    &.accordion-disabled {
        opacity: .5;

        .accordion-header-wrapper {
            cursor: default;
        }

        .accordion-header-caret {
        
            &::before {
                opacity: .5;
            }

        }

    }

}

/* Header */

.accordion-header-wrapper {
    display: flex;
    justify-content: space-between;
    align-items: flex-start;
    background: @color-text-main;
    color: @color-text-negative;
    position: relative;
    z-index: 1;
    cursor: pointer;

    .accordion-open & {
        border-radius: @base-border-radius @base-border-radius 0 0;
        transition: border-radius .2s ease-in-out;
    }

    .accordion-closed & {
        border-radius: @base-border-radius;
        transition: border-radius .2s ease-in-out;
    }

}

@accordion-caret-wrapper-width: @space-xlarge;

.accordion-header-caret-wrapper {
    padding: 0 @accordion-padding 0 @space-small;
    width: @accordion-caret-wrapper-width;
    position: relative;
    box-sizing: border-box;
}

@accordion-caret-size: 5px;

.accordion-header-caret {
        
    &::before {
        position: absolute;
        top: @accordion-padding + @accordion-caret-size;
        left: (@accordion-caret-wrapper-width / 2) - @accordion-caret-size;
        width: 0; 
        height: 0; 
        content: "";
        border-left: @accordion-caret-size solid transparent;
        border-right: @accordion-caret-size solid transparent;    
        border-top: @accordion-caret-size solid @color-text-negative;
        transform: rotate(0deg);
        transition: transform .2s ease-in-out;
    }

    .accordion-closed & {

        &::before {
            transform: rotate(180deg);
            transition: transform .2s ease-in-out;
        }

    }

}

.accordion-header-title {
    padding: @accordion-padding 0 @accordion-padding @accordion-padding;
    position: relative;
    z-index: 2;
}

/* Content */

.accordion-content-wrapper {
    overflow: hidden;
    max-height: 1000vh;
    transition: max-height 1s ease-in-out;

    .accordion-closed & {
        max-height: 0;
        transition: max-height .2s ease-in-out;
    }

    .accordion-open & {
        border-radius: 0 0 @base-border-radius @base-border-radius;
    }

}

.accordion-content {
    padding: @accordion-padding;
}
