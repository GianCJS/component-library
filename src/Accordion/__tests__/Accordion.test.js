import React from 'react';
import { mount, shallow } from 'enzyme';
import Accordion from '../index';

describe('Accordion Unit Test', () => {
    let props;
    let shallowedAccordion;

    const getAccordion = () => {
        if (!shallowedAccordion) {
            shallowedAccordion = shallow(<Accordion {...props} />);
        }
        return shallowedAccordion;
    };

    const mountedAccordion = () => mount(<Accordion {...props} />);

    beforeEach(() => {
        props = {
            header    : '',
            content   : '',
            open      : false,
            className : '',
            disabled  : false,
        };
        shallowedAccordion = undefined;
    });
    test('Component should render without crashing', () => {
        expect(getAccordion()).toBeDefined();
    });
    test('Component should render correctly', () => {
        expect(mountedAccordion()).toMatchSnapshot();
    });
    test('Should exists accordion-closed class when open is FALSE', () => {
        const wrapper = getAccordion();
        expect(wrapper.find('.accordion-closed').exists()).toBe(true);
    });
    test('Should exists accordion-open class when click in header-wrapper', () => {
        const wrapper = mountedAccordion();
        wrapper.find('.accordion-header-wrapper').simulate('click');
        expect(wrapper.find('.accordion-open').exists()).toBe(true);
    });
    test('Should exists accordion-disabled and accordion-closed class when disabled prop is TRUE', () => {
        props = {
            ...props,
            disabled: true,
        };
        const wrapper = getAccordion();
        expect(wrapper.find('.accordion-disabled').exists()).toBe(true);
        expect(wrapper.find('.accordion-closed').exists()).toBe(true);
    });
});
