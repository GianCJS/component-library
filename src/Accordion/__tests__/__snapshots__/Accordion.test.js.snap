// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Unit Test Component should render correctly 1`] = `
<Accordion
  card={true}
  className=""
  content=""
  disabled={false}
  header=""
  open={false}
>
  <div
    className="accordion-main-wrapper
        accordion-closed
        
        accordion-card
        
    "
  >
    <div
      className="accordion-header-wrapper"
      onClick={[Function]}
    >
      <div
        className="accordion-header-title"
      />
      <div
        className="accordion-header-caret-wrapper"
      >
        <span
          className="accordion-header-caret"
        />
      </div>
    </div>
    <div
      className="accordion-content-wrapper"
    >
      <div
        className="accordion-content"
      />
    </div>
  </div>
</Accordion>
`;
